File: misc-problems/editDistance.js
const EMPTY_STRING_LAST_INDEX = -1;

const deleteLastChar = (string) => {
  const lastIdx = string.length - 1;
  return string.substring(0, lastIdx);
};

const minDistanceBetweenWords = (
  word1,
  word2,
  word1LastIdx,
  word2LastIdx,
  minDistanceCache,
) => {
  const key = `${word1LastIdx}-${word2LastIdx}`;

  const keyInCache = minDistanceCache.hasOwnProperty(key);

  if (keyInCache) return minDistanceCache[key];

  if (word1LastIdx === EMPTY_STRING_LAST_INDEX) return word2LastIdx + 1;
  if (word2LastIdx === EMPTY_STRING_LAST_INDEX) return word1LastIdx + 1;

  const lastChar1 = word1[word1LastIdx];
  const lastChar2 = word2[word2LastIdx];

  if (lastChar1 === lastChar2) {
    minDistanceCache[key] = minDistanceBetweenWords(
      word1,
      word2,
      word1LastIdx - 1,
      word2LastIdx - 1,
      minDistanceCache,
    );

    return minDistanceCache[key];
  }

  const minDistanceIfDeleteFrom1 = minDistanceBetweenWords(
    word1,
    word2,
    word1LastIdx - 1,
    word2LastIdx,
    minDistanceCache,
  );
  const minDistanceIfDeleteFrom2 = minDistanceBetweenWords(
    word1,
    word2,
    word1LastIdx,
    word2LastIdx - 1,
    minDistanceCache,
  );
  const minDistanceIfDeleteFromBoth = minDistanceBetweenWords(
    word1,
    word2,
    word1LastIdx - 1,
    word2LastIdx - 1,
    minDistanceCache,
  );

  minDistanceCache[key] =
    Math.min(
      minDistanceIfDeleteFrom1,
      minDistanceIfDeleteFrom2,
      minDistanceIfDeleteFromBoth,
    ) + 1;

  return minDistanceCache[key];
};

const minDistance = (word1, word2) => {
  const minDistanceCache = {};

  const word1LastIdx = word1.length - 1;
  const word2LastIdx = word2.length - 1;

  return minDistanceBetweenWords(
    word1,
    word2,
    word1LastIdx,
    word2LastIdx,
    minDistanceCache,
  );
};

File: solutions/python/dp/2d/edit_distance.py
EMPTY_STRING_LAST_INDEX = -1


class Solution:
    def min_distance_between_words(self, word1, word2, word1_last_idx, word2_last_idx, min_distance_cache):
        key = f"{word1_last_idx}-{word2_last_idx}"

        if key in min_distance_cache:
            return min_distance_cache[key]

        if word1_last_idx == EMPTY_STRING_LAST_INDEX:
            return word2_last_idx + 1

        if word2_last_idx == EMPTY_STRING_LAST_INDEX:
            return word1_last_idx + 1

        last_char_1 = word1[word1_last_idx]
        last_char_2 = word2[word2_last_idx]

        if last_char_1 == last_char_2:
            min_distance_cache[key] = self.min_distance_between_words(
                word1,
                word2,
                word1_last_idx - 1,
                word2_last_idx - 1,
                min_distance_cache,
            )

            return min_distance_cache[key]

        min_distance_with_deletion = (
            1
            + self.min_distance_between_words(
                word1,
                word2,
                word1_last_idx - 1,
                word2_last_idx,
                min_distance_cache,
            )
        )

        min_distance_with_insertion = (
            1
            + self.min_distance_between_words(
                word1,
                word2,
                word1_last_idx,
                word2_last_idx - 1,
                min_distance_cache,
            )
        )

        min_distance_with_replacement = (
            1
            + self.min_distance_between_words(
                word1,
                word2,
                word1_last_idx - 1,
                word2_last_idx - 1,
                min_distance_cache,
            )
        )

        min_distance_cache[key] = min(
            min_distance_with_deletion,
            min_distance_with_insertion,
            min_distance_with_replacement,
        )

        return min_distance_cache[key]

    def minDistance(self, word1: str, word2: str) -> int:
        min_distance_cache = {}

        word_1_last_idx = len(word1) - 1
        word_2_last_idx = len(word2) - 1

        return self.min_distance_between_words(
            word1,
            word2,
            word_1_last_idx,
            word_2_last_idx,
            min_distance_cache,
        )
